<a id="1"></a>
<div style="padding:20px;
            color:#F0F3FF;
            margin:10;
            font-size:200%;
            text-align:left;
            display:fill;
            border-radius:15px;
            background-color:#211951;
            overflow:hidden;
            font-weight:900">
            <span style='color:#15F5BA'></span>Pandas : Data Manipulation
</div>
<br>Author : <i>Hanum Amaliyati Hayati</i>

# <b>A. <span style='color:#3A6D8C'>Introduction</span></b>

Sebelum sebuah data dapat dianalisis atau digunakan untuk mengambil keputusan kebijakan tentu data tersebut perlu untuk diekstraksi terlebih dahulu. <b>Ekstraksi data</b> adalah sebuah proses pengumpulan atau pengambilan jenis data yang berbeda dari berbagai sumber, banyak di antaranya mungkin tidak terorganisir dengan baik atau tidak terstruktur sama sekali. Proses ini memungkinkan untuk mengkonsolidasikan, memproses, dan menyempurnakan data sehingga nantinya dapat disimpan di lokasi terpusat untuk diubah atau diproses lebih lanjut.

Python sebagai salah satu <i>tools</i> yang <i>powerful</i> untuk pengolahan data menyediakan berbagai <i>library</i> untuk mendukung proses ekstraksi data. Salah satu library paling banyak digunakan adalah <b>pandas</b>.

Pandas memungkinkan untuk membaca data dari berbagai jenis ekstensi seperti data yang bersumber dari database (sql), <i>excel</i> (<b>.xls</b> atau <b>.xlsx</b>), <i>separated value file</i> (<b>.csv</b>, <b>.tsv</b>, dsb) lalu merubahnya menjadi satu format struktur data yang sama yang disebut <b>DataFrame</b>. DataFrame inilah yang nantinya dapat kita lakukan proses <b>Data Manipulation</b><br><br>

<p align="center">
<img src="https://raw.githubusercontent.com/bachtiyarma/Material/refs/heads/main/Image/Materi-Python/Pandas-DataFrame.png" width="600" height="300">
<p align="center"><b>Gambar 1. Proses Ekstraksi Data dengan <i>pandas</i></b></p>
<br><br>
<p align="left"><sup>docs : <i><a href="https://shorturl.at/xyZ69">https://pandas.pydata.org/docs/user_guide/io.html</i></sup></a></p>


## <b>A.1. <span style='color:#6A9AB0'>First Step : Pandas</span></b>  
Untuk dapat menggunakan library pandas pada python, perlu untuk mengimpor library tersebut dengan cara mengetikan sintaks berikut

<p align="center"><code><b>import pandas as pd</b></code></p>

lalu jalankan (<i>run</i>) program.
<br><br>
<p align="left"><sup>docs : <i><a href="https://shorturl.at/kzG06">https://pandas.pydata.org/docs/getting_started/install.html</i></sup></a></p>

###<b><span style='color:#55679C'>Quest A.1</span></b>
<i>Ketikan kode `import pandas as pd` pada prompt dibawah ini lalu jalankan!</i>

In [74]:
# Answer A.1
import pandas as pd

# <b>B. <span style='color:#3A6D8C'>Struktur Data Pandas</span></b>

* Terdapat dua struktur data utama pada library pandas yang disebut dengan Series dan DataFrame.
<p align="center">
<img src="https://datagy.io/wp-content/uploads/2022/12/Understanding-Pandas-Series-and-DataFrames-1024x576.png" width="550" height="289">
<p align="center"><b>Gambar 2. Series VS DataFrame</b></p>
<br>

* Gabungan beberapa Series yang mempunyai banyak baris yang sama dapat membentuk sebuah DataFrame atau berlaku sebaliknya pada sebuah DataFrame dapat dipecah (di-<i>slicing</i> satu kolom) membentuk sebuah Series.<br><br>

<ul><p align="left"><sup>docs : <i><a href="https://shorturl.at/dezQY">http://pandas.pydata.org/docs/getting_started/intro_tutorials/01_table_oriented.html</i></sup></a></p></ul>

## <b>B.1. <span style='color:#6A9AB0'>Series</span></b>  

* <a href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html">Series</a> merupakan kolom dari sebuah tabel yang bersifat homogen atau mempunyai elemen dengan tipe data yang sejenis.

* Untuk membuat Series gunakan sintaks berikut

<p align="center"><code><b>pd.Series(data)</b></code></p>

<ul>
  data dapat berupa <i>array-like</i>, <i>Iterable</i>, <i>dict</i>, atau <i>scalar</i>
</ul>
<br>
<ul><p align="left"><sup>docs : <i><a href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html">https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html</i></sup></a></p></ul>



###<b><span style='color:#55679C'>Quest B.1.1</span></b>
<i>Hanya Singapura dan Indonesia yang saat ini telah menetapkan pajak karbon (carbon pricing) di Asia Tenggara. (source : https://accept.aseanenergy.org/progress-of-carbon-pricing-in-asean)

Buatlah sebuah Series yang disimpan pada variabel dengan nama `country_carbon_pricing` untuk menampung nama negara yang telah menetapkan pajak karbon di Asia Tenggara, lalu tampilkan outputnya!</i>

In [75]:
# Answer B.1.1
countries = ['Singapura', 'Indonesia']
country_carbon_pricing = pd.Series(countries)
print(country_carbon_pricing)

0    Singapura
1    Indonesia
dtype: object


In [76]:
#cek rype data
country = ['Singapura', 'Indonesia']
type(country)

list

List =  lebih flesibel untuk tipe datanya

Array = tipe datanya harus sama

###<b><span style='color:#55679C'>Quest B.1.2</span></b>
<i>Dari data emisi karbon yang tercatat di Indonesia, akan diberi detail data sebagai berikut

* Tahun 2015 dihasilkan 491.46 Mton CO2
* Tahun 2016 dihasilkan 488.40 Mton CO2
* Tahun 2017 dihasilkan 511.33 Mton CO2
* Tahun 2018 dihasilkan 579.23 Mton CO2
* Tahun 2019 dihasilkan 625.66 Mton CO2

<i>Source: Joint Research Centre of European Union</i>

Dari data tersebut buatlah sebuah Series yang disimpan pada variabel dengan nama `emisi_co2_indonesia` lalu tampilkan outputnya!</i>

In [77]:
data_emisi = {
    "Tahun 2015": 491.46,
    "Tahun 2016": 488.40,
    "Tahun 2017": 511.33,
    "Tahun 2018": 579.23,
    "Tahun 2019": 625.66}

emisi_co2_indonesia = pd.Series(data_emisi)
print(emisi_co2_indonesia)

Tahun 2015    491.46
Tahun 2016    488.40
Tahun 2017    511.33
Tahun 2018    579.23
Tahun 2019    625.66
dtype: float64


## <b>B.2. <span style='color:#6A9AB0'>DataFrame</span></b>  

* <a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html#pandas.DataFrame">DataFrame</a> adalah struktur data 2 dimensi yang terdiri dari baris dan kolom (seperti tabel) dan dapat menyimpan data dari berbagai jenis (termasuk karakter, bilangan bulat, bilangan real, data kategorikal, dan lainnya).

<p align="center">
<img src="https://www.w3resource.com/w3r_images/pandas-data-frame.svg" width="600" height="300">
<p align="center"><b>Gambar 2. DataFrame Pandas</b></p>
<br>

* <a href="https://pandas.pydata.org/docs/reference/api/pandas.Series.html#pandas.Series">Series</a> merupakan kolom dari sebuah tabel yang bersifat homogen atau mempunyai elemen dengan tipe data yang sejenis.


<p align="center">
<img src="https://datagy.io/wp-content/uploads/2022/12/Understanding-Pandas-Series-and-DataFrames-1024x576.png" width="550" height="300">
<p align="center"><b>Gambar 3. Series VS DataFrame</b></p>
<br>

* Gabungan beberapa Series yang mempunyai banyak baris yang sama dapat membentuk sebuah DataFrame atau berlaku sebaliknya pada sebuah DataFrame dapat dipecah (di-<i>slicing</i> satu kolom) membentuk sebuah Series.<br><br>

* Untuk membuat sebuah DataFrame secara manual, dapat menggunakan sintaks
<p align="center"><code><b>pd.DataFrame(data)</b></code></p>
<ul>data inputan dapat berupa ndarray (structured atau homogeneous), iterable atau dict</ul>
<br>
<ul><p align="left"><sup>docs : <i><a href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html">https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html</i></sup></a></p></ul>



###<b><span style='color:#55679C'>Quest B.2</span></b>
<i>Diberikan data Air Quality Index (AQI) dan partikel udara yang memiliki diameter kurang dari 2,5 mikrometer (PM2.5) di Jakarta (source : <a href="https://www.iqair.com/id/indonesia/jakarta?srsltid=AfmBOoqcwssGzLEzHMeGJwjaU-oi7wh1JeUqR-wBxicstXuvUN_i5UwT">https://www.iqair.com/id/indonesia/jakarta?srsltid=AfmBOoqcwssGzLEzHMeGJwjaU-oi7wh1JeUqR-wBxicstXuvUN_i5UwT</a>) sebagai berikut :</i>
<br><br>
<table>
    <tr>
        <th>Tanggal</th>
        <th>PM2.5</th>
        <th>AQI</th>
    </tr>
    <tr>
        <td>2024-09-21</td>
        <td>70.7</td>
        <td>162</td>
    </tr>
    <tr>
        <td>2024-09-22</td>
        <td>58.6</td>
        <td>153</td>
    </tr>
    <tr>
        <td>2024-09-23</td>
        <td>52.3</td>
        <td>142</td>
    </tr>
</table>
<br>

<i>Dari data diatas, buatkan secara manual DataFrame-nya dengan pandas!</i>



In [78]:
# Answer B.2
data_aqi={
    'Tanggal': ['2024-09-21', '2024-09-22', '2024-09-23'],
    'PM2.5' : [70.7, 58.6, 52.3],
    'AQI' : [162, 153, 142]}

df =  pd.DataFrame(data_aqi)
display(df)

Tanggal  PM2.5  AQI
0  2024-09-21   70.7  162
1  2024-09-22   58.6  153
2  2024-09-23   52.3  142

## <b>B.3. <span style='color:#6A9AB0'>Parts Of DataFrame</span></b>

Untuk lebih memahami objek DataFrame, perlu diketahui bahwa DataFrame terdiri dari tiga komponen sebagai atribut berikut :

<ul>1️⃣ .values : Nilai - nilai yang tertera pada DataFrame</ul>
<ul>2️⃣ .columns : Nama kolom yang ada pada DataFrame</ul>
<ul>3️⃣ .index : Nama atau nomor index pada DataFrame</ul>



###<b><span style='color:#55679C'>Quest B.3.1</span></b>

<i>Pada DataFrame yang telah dibuat pada Quest B.2, coba buatkan list nama kolomnya!</i>


In [79]:
# Answer B.3.1
print(df.columns)

Index(['Tanggal', 'PM2.5', 'AQI'], dtype='object')


###<b><span style='color:#55679C'>Quest B.3.2</span></b>

<i>Pada DataFrame yang telah dibuat pada Quest B.2, coba akses setiap nilainya!</i>


In [80]:
# Answer B.3.2
print(df.values)

[['2024-09-21' 70.7 162]
 ['2024-09-22' 58.6 153]
 ['2024-09-23' 52.3 142]]


# <b>C. <span style='color:#3A6D8C'>Data Manipulation</span></b>

Data Manipulation adalah sebuah metode pengorganisasian data agar lebih mudah dibaca, terstruktur dan agar lebih mudah diolah untuk kebutuhan lain kedepannya. Beberapa hal yang biasa dilakukan saat data manipulation adalah sebagai berikut :


<ul>1️⃣ Column selections</ul>
<ul>2️⃣ Filter the data</ul>
<ul>3️⃣ Create a new column or change it</ul>
<ul>4️⃣ Drop a column</ul>
<ul>5️⃣ Apply custom functions</ul>
<ul>6️⃣ Join / Merge the data</ul>
<ul>7️⃣ etc</ul>


## <b>C.1. <span style='color:#6A9AB0'>Data Extraction</span></b>

* Dalam konteks data extraction, Pandas dapat digunakan untuk membaca data dari berbagai sumber, seperti file CSV, Excel, SQL database, dan lainnya, dan mengubahnya menjadi DataFrames untuk analisis lebih lanjut.

* Beberapa fungsi dan metode Pandas yang umum digunakan untuk data extraction adalah sebagai berikut:

<ul>
<code><a href="https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html">pd.read_csv()</a></code>: Digunakan untuk membaca data dari file CSV dan mengonversinya menjadi DataFrame.

<code><a href="https://pandas.pydata.org/docs/reference/api/pandas.read_excel.html">pd.read_excel()</a></code>: Membaca data dari file Excel dan membuat DataFrame dari lembar kerja atau tabel tertentu.

<code><a href="https://pandas.pydata.org/docs/reference/api/pandas.read_sql.html">pd.read_sql()</a></code>: Mengambil data dari basis data SQL menggunakan SQL query dan menyimpannya dalam bentuk DataFrame.
</ul>


###<b><span style='color:#55679C'>Quest C.1.</span></b>

<i>Untuk melakukan proses ekstraksi data, lakukan instruksi berikut :     

1.   Persiapkan dataset WorldSustainabilityDataset.csv
(dataset : https://drive.google.com/file/d/1v0r3fqlT20NxhjDuDZCiswih9jTGOHwd/view?usp=sharing)
2.   Upload ke google colab dengan cara pilih File pada bagian kiri laman (icon 📁), kemudian klik kanan dan pilih upload
3.   Lakukan ekstraksi data csv (.csv) dengan library pandas, assign hasilnya ke sebuah variabel bernama <code>data_world_sust</code>
4.   Tampilkan hasilnya</i>






In [81]:
# Answer C.1.
data = '/content/WorldSustainabilityDataset.csv'
data_world_sust = pd.read_csv(data)
display(data_world_sust.head())

Country Name Country Code  Year  \
0        Aruba          ABW  2000   
1        Aruba          ABW  2001   
2        Aruba          ABW  2002   
3        Aruba          ABW  2003   
4        Aruba          ABW  2004   

   Access to electricity (% of population) - EG.ELC.ACCS.ZS  \
0                                          91.660398          
1                                         100.000000          
2                                         100.000000          
3                                         100.000000          
4                                         100.000000          

   Adjusted net national income per capita (annual % growth) - NY.ADJ.NNTY.PC.KD.ZG  \
0                                                NaN                                  
1                                                NaN                                  
2                                                NaN                                  
3                                                NaN                                  
4                                                NaN                                  

   Adjusted net savings, excluding particulate emission damage (% of GNI) - NY.ADJ.SVNX.GN.ZS  \
0                                          15.392211                                            
1                                          11.491930                                            
2                                           3.893620                                            
3                                           6.713863                                            
4                                           9.801576                                            

   Adjusted savings: carbon dioxide damage (% of GNI) - NY.ADJ.DCO2.GN.ZS  \
0                                           2.389134                        
1                                           2.509046                        
2                                           2.721385                        
3                                           2.733676                        
4                                           2.678363                        

   Adjusted savings: natural resources depletion (% of GNI) - NY.ADJ.DRES.GN.ZS  \
0                                           0.000190                              
1                                           0.000214                              
2                                           0.000223                              
3                                           0.000243                              
4                                           0.000256                              

   Adjusted savings: net forest depletion (% of GNI) - NY.ADJ.DFOR.GN.ZS  \
0                                           0.000190                       
1                                           0.000214                       
2                                           0.000223                       
3                                           0.000243                       
4                                           0.000256                       

   Adjusted savings: particulate emission damage (% of GNI) - NY.ADJ.DPEM.GN.ZS  \
0                                                NaN                              
1                                                NaN                              
2                                                NaN                              
3                                                NaN                              
4                                                NaN                              

   ...  Gini index (World Bank estimate) - SI.POV.GINI  \
0  ...                                             NaN   
1  ...                                             NaN   
2  ...                                             NaN   
3  ...                                             NaN   
4  ...                                             NaN   

   Income Classification (W

## <b>C.2. <span style='color:#6A9AB0'>Inspecting Data Information</span></b>

* Untuk dapat menginspeksi informasi umum pada data dapat menggunakan sintaks berikut

<p align="center"><code><b>DataFrame.info()</b></code></p>

<ul>Ketika dijalankan, sintaks tersebut akan menghasilkan output informasi umum pada data seperti banyak kolom, nama kolom beserta tipe data dan jumlah data tak kosong (<i>non-null value</i>)</ul><br>

<ul><p align="left"><sup>docs : <i><a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html">https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html</i></sup></a></p></ul>


### <b><span style='color:#55679C'>Quest C.2</span></b>

<i>Inspeksi informasi umum pada data yang telah diekstraksi dan telah disimpan pada variabel <code>data_world_sust</code>. Setelah itu coba uraikan informasi apa yang bisa didapatkan setelah melakukan inspeksi informasi umum pada <code>data_world_sust</code>?</i>

In [82]:
# Answer C.2
data_world_sust.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3287 entries, 0 to 3286
Data columns (total 54 columns):
 #   Column                                                                                        Non-Null Count  Dtype  
---  ------                                                                                        --------------  -----  
 0   Country Name                                                                                  3287 non-null   object 
 1   Country Code                                                                                  3287 non-null   object 
 2   Year                                                                                          3287 non-null   int64  
 3   Access to electricity (% of population) - EG.ELC.ACCS.ZS                                      3193 non-null   float64
 4   Adjusted net national income per capita (annual % growth) - NY.ADJ.NNTY.PC.KD.ZG              2566 non-null   float64
 5   Adjusted net savings, exclu

## <b>C.3. <span style='color:#6A9AB0'>Data Selection</span></b>

* Untuk memilih kolom tertentu pada sebuah DataFrame dapat menggunakan sintaks berikut

<p align="center"><code><b>DataFrame.loc[baris , nama_kolom]</b></code></p>

* <code>baris</code>
<ul>
  <li>Gunakan simbol ( : ) untuk menampilkan semua data</li>
  <li>Gunakan simbol (m : n) untuk menampilkan data dari index ke - m sampai dengan index ke - n</li>
</ul>

* <code>nama_kolom</code>
<ul>
  <li>Gunakan string untuk mengakses satu kolom saja (<i>return : Series</i>)</li>
  <li>Gunakan list untuk mengakses lebih dari satu kolom  (<i>return : DataFrame</i>)</li>
</ul>


###<b><span style='color:#55679C'>Quest C.3.1</span></b>

<i>Pada DataFrame <code>data_world_sust</code>, simpan 18 data teratas ke dalam variabel **year** untuk kolom **Year** saja. Kemudian tampilkan data tersebut!</i>




In [83]:
year = data_world_sust.loc[0 : 18, 'Year']
display(year)

0     2000
1     2001
2     2002
3     2003
4     2004
5     2005
6     2006
7     2007
8     2008
9     2009
10    2010
11    2011
12    2012
13    2013
14    2014
15    2015
16    2016
17    2017
18    2018
Name: Year, dtype: int64

###<b><span style='color:#55679C'>Quest C.3.2</span></b>

<i>Pada DataFrame <code>data_world_sust</code>, Tampilkan 18 data teratas untuk kolom
* Country Name
* Year
* Access to electricity (% of population) - EG.ELC.ACCS.ZS</i>

In [84]:
# Answer C.3.2
data_sort = data_world_sust.loc[0:18, ['Country Name', 'Year','Access to electricity (% of population) - EG.ELC.ACCS.ZS']]
display(data_sort)

Country Name  Year  \
0         Aruba  2000   
1         Aruba  2001   
2         Aruba  2002   
3         Aruba  2003   
4         Aruba  2004   
5         Aruba  2005   
6         Aruba  2006   
7         Aruba  2007   
8         Aruba  2008   
9         Aruba  2009   
10        Aruba  2010   
11        Aruba  2011   
12        Aruba  2012   
13        Aruba  2013   
14        Aruba  2014   
15        Aruba  2015   
16        Aruba  2016   
17        Aruba  2017   
18        Aruba  2018   

    Access to electricity (% of population) - EG.ELC.ACCS.ZS  
0                                           91.660398         
1                                          100.000000         
2                                          100.000000         
3                                          100.000000         
4                                          100.000000         
5                                          100.000000         
6                                          100.000000         
7                                          100.000000         
8                                          100.000000         
9                                          100.000000         
10                                          93.356292         
11                                         100.000000         
12                                         100.000000         
13                                         100.000000         
14                                         100.000000         
15                                         100.000000         
16                                         100.000000         
17                                         100.000000         
18                                         100.000000

## <b>C.4. <span style='color:#6A9AB0'>Conditions</span></b>

* Salah satu teknik paling sederhana untuk menemukan bagian mana dari kumpulan data yang menarik adalah menemukan subset baris yang cocok dengan beberapa kriteria untuk difilter atau dipilih. Untuk melakukan hal ini cukup mudah yakni dengan menuliskan kondisi seperti berikut

<p align="center"><code><b>DataFrame.loc[conditions]</b></code></p>

* Conditions dapat dibentuk dari 1 atau lebih kriteria boolean dan dihubungkan dengan operator logika and (&), or (|) atau not (~)

<br>

docs : <i>https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.loc.html</i>

###<b><span style='color:#55679C'>Quest C.4.1</span></b>

<i>Pada DataFrame <code>data_world_sust</code>, tampilkan semua data yang Country Name-nya Indonesia</i>

In [85]:
# Answer C.4.1
data_Indonesia = data_world_sust.loc[data_world_sust['Country Name'] == 'Indonesia']
display(data_Indonesia)

Country Name Country Code  Year  \
1368    Indonesia          IDN  2000   
1369    Indonesia          IDN  2001   
1370    Indonesia          IDN  2002   
1371    Indonesia          IDN  2003   
1372    Indonesia          IDN  2004   
1373    Indonesia          IDN  2005   
1374    Indonesia          IDN  2006   
1375    Indonesia          IDN  2007   
1376    Indonesia          IDN  2008   
1377    Indonesia          IDN  2009   
1378    Indonesia          IDN  2010   
1379    Indonesia          IDN  2011   
1380    Indonesia          IDN  2012   
1381    Indonesia          IDN  2013   
1382    Indonesia          IDN  2014   
1383    Indonesia          IDN  2015   
1384    Indonesia          IDN  2016   
1385    Indonesia          IDN  2017   
1386    Indonesia          IDN  2018   

      Access to electricity (% of population) - EG.ELC.ACCS.ZS  \
1368                                          86.300000          
1369                                          86.260000          
1370                                          87.600000          
1371                                          87.940000          
1372                                          89.010000          
1373                                          85.421112          
1374                                          90.620000          
1375                                          91.100000          
1376                                          92.730000          
1377                                          93.550000          
1378                                          94.150000          
1379                                          94.830000          
1380                                          96.000000          
1381                                          96.464258          
1382                                          97.010000          
1383                                          97.537367          
1384                                          97.620000          
1385                                          98.140000          
1386                                          98.510000          

      Adjusted net national income per capita (annual % growth) - NY.ADJ.NNTY.PC.KD.ZG  \
1368                                         175.094481                                  
1369                                          29.740863                                  
1370                                        -128.869275                                  
1371                                        -446.094475                                  
1372                                         -88.418979                                  
1373                                         229.493078                                  
1374                                         -52.439762                                  
1375                                        -645.655290                                  
1376                                          -2.618162                                  
1377                                          22.017902                                  
1378                                           1.319571                                  
1379                                           4.228906                                  
1380                                           3.675770                                  
1381                                           4.185756                                  
1382                                           2.529367                                  
1383                                           5.124025                                  
1384                                           6.383631                                  
1385                                           4.928758                                  
1386                                           4.733578                                  

      Adjusted net savings, excluding particulate emission damage (% of GNI) - NY.

###<b><span style='color:#55679C'>Quest C.4.2</span></b>

<i>Pada DataFrame <code>data_world_sust</code>, tampilkan semua data yang Country Name-nya Indonesia namun hanya untuk tahun 2012 keatas saja</i>

In [86]:
# Answer C.4.2
Indonesia_2012= data_world_sust.loc[(data_world_sust['Country Name'] == 'Indonesia') & (data_world_sust['Year'] >= 2012)]
display(Indonesia_2012)

Country Name Country Code  Year  \
1380    Indonesia          IDN  2012   
1381    Indonesia          IDN  2013   
1382    Indonesia          IDN  2014   
1383    Indonesia          IDN  2015   
1384    Indonesia          IDN  2016   
1385    Indonesia          IDN  2017   
1386    Indonesia          IDN  2018   

      Access to electricity (% of population) - EG.ELC.ACCS.ZS  \
1380                                          96.000000          
1381                                          96.464258          
1382                                          97.010000          
1383                                          97.537367          
1384                                          97.620000          
1385                                          98.140000          
1386                                          98.510000          

      Adjusted net national income per capita (annual % growth) - NY.ADJ.NNTY.PC.KD.ZG  \
1380                                           3.675770                                  
1381                                           4.185756                                  
1382                                           2.529367                                  
1383                                           5.124025                                  
1384                                           6.383631                                  
1385                                           4.928758                                  
1386                                           4.733578                                  

      Adjusted net savings, excluding particulate emission damage (% of GNI) - NY.ADJ.SVNX.GN.ZS  \
1380                                          11.329746                                            
1381                                          11.232642                                            
1382                                          11.101481                                            
1383                                          11.798671                                            
1384                                          13.691120                                            
1385                                          14.720104                                            
1386                                          15.333506                                            

      Adjusted savings: carbon dioxide damage (% of GNI) - NY.ADJ.DCO2.GN.ZS  \
1380                                           1.844120                        
1381                                           1.514861                        
1382                                           1.445983                        
1383                                           2.141863                        
1384                                           2.243018                        
1385                                           2.181155                        
1386                                           2.273700                        

      Adjusted savings: natural resources depletion (% of GNI) - NY.ADJ.DRES.GN.ZS  \
1380                                           3.709363                              
1381                                           3.308081                              
1382                                           2.773960                              
1383                                           1.650662                              
1384                                           1.383614                              
1385                                           1.686335                              
1386                                           2.521513                              

      Adjusted savings: net forest depletion (% of GNI) - NY.ADJ.DFOR.GN.ZS  \
1380                                                0.0                       
1381                                                0.0                       
1382                                                0.0           

###<b><span style='color:#55679C'>Quest C.4.3</span></b>

<i>Pada DataFrame <code>data_world_sust</code>, tampilkan kolom 'Country Name', 'GDP per capita (current US$) - NY.GDP.PCAP.CD', 'Inflation, consumer prices (annual %) - FP.CPI.TOTL.ZG' yang Country Name-nya Indonesia</i>

In [87]:
Indonesia= data_world_sust.loc[
    (data_world_sust['Country Name'] == 'Indonesia'),
    ['Country Name', 'GDP per capita (current US$) - NY.GDP.PCAP.CD',
    'Inflation, consumer prices (annual %) - FP.CPI.TOTL.ZG']]

display(Indonesia)

Country Name  GDP per capita (current US$) - NY.GDP.PCAP.CD  \
1368    Indonesia                                     780.190205   
1369    Indonesia                                     748.257609   
1370    Indonesia                                     900.177588   
1371    Indonesia                                    1065.648520   
1372    Indonesia                                    1150.261367   
1373    Indonesia                                    1263.287332   
1374    Indonesia                                    1589.801489   
1375    Indonesia                                    1860.002811   
1376    Indonesia                                    2166.854231   
1377    Indonesia                                    2261.247309   
1378    Indonesia                                    3122.362673   
1379    Indonesia                                    3643.047176   
1380    Indonesia                                    3694.359340   
1381    Indonesia                                    3623.927241   
1382    Indonesia                                    3491.637491   
1383    Indonesia                                    3331.695115   
1384    Indonesia                                    3562.816334   
1385    Indonesia                                    3837.578024   
1386    Indonesia                                    3893.859578   

      Inflation, consumer prices (annual %) - FP.CPI.TOTL.ZG  
1368                                           3.688619       
1369                                          11.500115       
1370                                          11.900118       
1371                                           6.757317       
1372                                           6.064060       
1373                                          10.453198       
1374                                          13.108672       
1375                                           6.406563       
1376                                          10.226665       
1377                                           4.386416       
1378                                           5.134204       
1379                                           5.356048       
1380                                           4.279500       
1381                                           6.412513       
1382                                           6.394925       
1383                                           6.363121       
1384                                           3.525805       
1385                                           3.808798       
1386                                           3.198346

###<b><span style='color:#55679C'>Quest C.4.4</span></b>

<i>Pada DataFrame <code>data_world_sust</code>, Pada tahun 2018, bandingkan GDP per capita pada negara Indonesia, Singapura, Malaysia dan Brunei Darussalam. Tampilkan kolom 'Country Name', 'Year', 'GDP per capita (current US$) - NY.GDP.PCAP.CD' saja </i>

In [88]:
# Answer C.4.4
asia = data_world_sust.loc[
    (data_world_sust['Year'] == 2018) &
    (data_world_sust['Country Name'].isin(['Indonesia', 'Malaysia', 'Singapore', 'Brunei Darussalam'])), #isin = IN
    ['Country Name', 'GDP per capita (current US$) - NY.GDP.PCAP.CD']]
display(asia)

Country Name  GDP per capita (current US$) - NY.GDP.PCAP.CD
474   Brunei Darussalam                                   31628.476260
1386          Indonesia                                    3893.859578
2184           Malaysia                                   11377.654200
2659          Singapore                                   66679.046490

## <b>C.5. <span style='color:#6A9AB0'>Sorting Row</span></b>

Menemukan data yang menarik dalam sebuah DataFrame seringkali lebih mudah jika dilakukan pengurutan baris datanya. Pengurutan dibagi menjadi dua yakni pengurutan secara `Ascending` (pengurutan naik A → Z) dan juga pengurutan secara `Descending` (pengurutan turun Z → A). Eksekusi sintaks berikut :

<p align="center"><code><b>DataFrame.sort_values(by, ascending=True, ignore_index=False)</b></code></p>

Jika ingin mengurutkan DataFrame pandas dengan memperhatikan argumen

<ul>
<li><b><i>by</i></b> sebagai satu atau lebih kolom yang ingin dijadikan patokan pengurutan</li>

<li><b><i>ascending</i></b> dapat diisikan True jika ingin terurut naik atau False jika ingin terurut turun (dafault : True)</li>

<li><b><i>ignore_index</i></b> dapat diisikan True jika ingin index masih terurut atau False jika indexnya masih dipertahankan sesuai data awal (dafault : False)</li>

</ul><br>

docs : <i>https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html</i>


###<b><span style='color:#55679C'>Quest C.5.1</span></b>

<i>Pada DataFrame <code>data_world_sust</code>, tampilkan semua data yang Country Name-nya Indonesia namun urutkan dari GDP dari yang terbesar sampai yang terkecil</i>

In [89]:
# Answer C.5.1
data_indo = (data_world_sust[data_world_sust['Country Name'] == 'Indonesia']
             .sort_values(by='GDP per capita (current US$) - NY.GDP.PCAP.CD', ascending = False, ignore_index = False))
display(data_indo)

Country Name Country Code  Year  \
1386    Indonesia          IDN  2018   
1385    Indonesia          IDN  2017   
1380    Indonesia          IDN  2012   
1379    Indonesia          IDN  2011   
1381    Indonesia          IDN  2013   
1384    Indonesia          IDN  2016   
1382    Indonesia          IDN  2014   
1383    Indonesia          IDN  2015   
1378    Indonesia          IDN  2010   
1377    Indonesia          IDN  2009   
1376    Indonesia          IDN  2008   
1375    Indonesia          IDN  2007   
1374    Indonesia          IDN  2006   
1373    Indonesia          IDN  2005   
1372    Indonesia          IDN  2004   
1371    Indonesia          IDN  2003   
1370    Indonesia          IDN  2002   
1368    Indonesia          IDN  2000   
1369    Indonesia          IDN  2001   

      Access to electricity (% of population) - EG.ELC.ACCS.ZS  \
1386                                          98.510000          
1385                                          98.140000          
1380                                          96.000000          
1379                                          94.830000          
1381                                          96.464258          
1384                                          97.620000          
1382                                          97.010000          
1383                                          97.537367          
1378                                          94.150000          
1377                                          93.550000          
1376                                          92.730000          
1375                                          91.100000          
1374                                          90.620000          
1373                                          85.421112          
1372                                          89.010000          
1371                                          87.940000          
1370                                          87.600000          
1368                                          86.300000          
1369                                          86.260000          

      Adjusted net national income per capita (annual % growth) - NY.ADJ.NNTY.PC.KD.ZG  \
1386                                           4.733578                                  
1385                                           4.928758                                  
1380                                           3.675770                                  
1379                                           4.228906                                  
1381                                           4.185756                                  
1384                                           6.383631                                  
1382                                           2.529367                                  
1383                                           5.124025                                  
1378                                           1.319571                                  
1377                                          22.017902                                  
1376                                          -2.618162                                  
1375                                        -645.655290                                  
1374                                         -52.439762                                  
1373                                         229.493078                                  
1372                                         -88.418979                                  
1371                                        -446.094475                                  
1370                                        -128.869275                                  
1368                                         175.094481                                  
1369                                          29.740863                                  

      Adjusted net savings, excluding particulate emission damage (% of GNI) - NY.

###<b><span style='color:#55679C'>Quest C.5.2</span></b>

<i>Pada DataFrame <code>data_world_sust</code>, pada tahun 2018, negara mana saja yang GDPnya masuk kedalam 10 terbaik? Tampilkan hanya kolom 'Country Name', 'Year', 'GDP per capita (current US$) - NY.GDP.PCAP.CD'  </i>

In [90]:
# Answer C.5.2
top10_GDP = (data_world_sust[data_world_sust['Year']== 2018]
            [['Country Name', 'Year', 'GDP per capita (current US$) - NY.GDP.PCAP.CD']]
             .sort_values(by='GDP per capita (current US$) - NY.GDP.PCAP.CD', ascending = False))
display(top10_GDP.head(10))

Country Name  Year  GDP per capita (current US$) - NY.GDP.PCAP.CD
1842        Luxembourg  2018                                   116597.29560
1880  Macao SAR, China  2018                                    87554.71386
569        Switzerland  2018                                    86429.50082
2298            Norway  2018                                    82267.80932
1424           Ireland  2018                                    79297.72920
1481           Iceland  2018                                    74348.36228
2659         Singapore  2018                                    66679.04649
2545             Qatar  2018                                    65907.94843
3134     United States  2018                                    63064.41841
873            Denmark  2018                                    61598.53670

###<b><span style='color:#55679C'>Quest C.5.3</span></b>

<i>Pada DataFrame <code>data_world_sust</code>, pada tahun 2016, negara mana saja yang Emisi karbon dioksida (CO2) berbasis produksi tahunannya masuk dalam 5 negara penghasil emisi CO2 terkecil? Tampilkan nama negara, tahun dan jumlah emisi karbonnya! </i>

In [91]:
# Answer C.5.3
top5_smallest_emitters_2016 = (data_world_sust[data_world_sust['Year']== 2016]
                               [['Country Name', 'Year', 'Annual production-based emissions of carbon dioxide (CO2), measured in million tonnes']]
                               .sort_values(by = 'Annual production-based emissions of carbon dioxide (CO2), measured in million tonnes', ascending = True))
display(top5_smallest_emitters_2016.head(5))

Country Name  Year  \
2980                           Tonga  2016   
3227                         Vanuatu  2016   
852                         Dominica  2016   
719                          Comoros  2016   
3170  St. Vincent and the Grenadines  2016   

      Annual production-based emissions of carbon dioxide (CO2), measured in million tonnes  
2980                                              0.128                                      
3227                                              0.147                                      
852                                               0.180                                      
719                                               0.202                                      
3170                                              0.253

###<b><span style='color:#55679C'>Quest C.5.4</span></b>

<i>Pada DataFrame <code>data_world_sust</code>, pada tahun 2018 di benua Asia, tampilkan 10 negara dengan produksi Emisi CO2 terbesar. Urutkan dari yang terbesar hingga terkecil</i>

In [92]:
#Answer C.5.4
top10_smaller_emitters_2018 = data_world_sust.loc[
    (data_world_sust['Year'] == 2018) &
    (data_world_sust['Continent'] == 'Asia'),
    ['Country Name', 'Year', 'Annual production-based emissions of carbon dioxide (CO2), measured in million tonnes']
].sort_values(by='Annual production-based emissions of carbon dioxide (CO2), measured in million tonnes', ascending=True)

# Menampilkan 10 negara dengan emisi CO2 terendah
display(top10_smaller_emitters_2018.head(10))


Country Name  Year  \
2963         Timor-Leste  2018   
1956            Maldives  2018   
493               Bhutan  2018   
1880    Macao SAR, China  2018   
2526  West Bank and Gaza  2018   
113              Armenia  2018   
2944          Tajikistan  2018   
474    Brunei Darussalam  2018   
1139             Georgia  2018   
1633     Kyrgyz Republic  2018   

      Annual production-based emissions of carbon dioxide (CO2), measured in million tonnes  
2963                                              0.520                                      
1956                                              1.565                                      
493                                               1.662                                      
1880                                              2.216                                      
2526                                              3.220                                      
113                                               6.297                                      
2944                                              7.473                                      
474                                               9.560                                      
1139                                              9.862                                      
1633                                             10.169

## <b>C.6. <span style='color:#6A9AB0'>Summarize Data</span></b>

Di pandas, method `DataFrame.describe()` digunakan untuk menghasilkan statistik deskriptif dari DataFrame. Metode ini sangat berguna untuk dengan cepat memahami statistik dasar dari data. Berikut adalah statistik utama yang disediakan :

* `Jumlah` : Jumlah nilai non-null dalam setiap kolom.
* `Rata-rata` : Rata-rata aritmetika dari setiap kolom.
* `Standar Deviasi` : Ukuran dispersi nilai-nilai sekitar rata-rata.
* `Minimum` : Nilai minimum dalam setiap kolom.
* `Kuartil 25 (Q1)` : Nilai di bawahnya 25% data berada.
* `Kuartil 50 (Median atau Q2)` : Nilai di bawahnya 50% data berada (yaitu nilai tengah).
* `Kuartil 75 (Q3)` : Nilai di bawahnya 75% data berada.
* `Maksimum` : Nilai maksimum dalam setiap kolom.

###<b><span style='color:#55679C'>Quest C.6</span></b>

<i>Pada DataFrame <code>data_world_sust</code>, hitung semua statistik deskriptif menggunakan .describe() pada kolom numerik!</i>

In [93]:
# Answer C.6
statistika_desk = data_world_sust.describe()
display(statistika_desk)

Year  Access to electricity (% of population) - EG.ELC.ACCS.ZS  \
count  3287.000000                                        3193.000000          
mean   2009.000000                                          79.300662          
std       5.478059                                          30.398748          
min    2000.000000                                           1.270180          
25%    2004.000000                                          62.900000          
50%    2009.000000                                          98.653969          
75%    2014.000000                                         100.000000          
max    2018.000000                                         100.000000          

       Adjusted net national income per capita (annual % growth) - NY.ADJ.NNTY.PC.KD.ZG  \
count                                        2566.000000                                  
mean                                            2.499857                                  
std                                            18.752649                                  
min                                          -645.655290                                  
25%                                            -0.246761                                  
50%                                             2.545392                                  
75%                                             5.545831                                  
max                                           229.493078                                  

       Adjusted net savings, excluding particulate emission damage (% of GNI) - NY.ADJ.SVNX.GN.ZS  \
count                                        2628.000000                                            
mean                                            8.870028                                            
std                                            12.277347                                            
min                                           -97.524776                                            
25%                                             3.339159                                            
50%                                             8.752619                                            
75%                                            15.290790                                            
max                                            45.354086                                            

       Adjusted savings: carbon dioxide damage (% of GNI) - NY.ADJ.DCO2.GN.ZS  \
count                                        3220.000000                        
mean                                            1.742674                        
std                                             1.957442                        
min                                             0.090292                        
25%                                             0.741310                        
50%                                             1.145867                        
75%                                             2.056927                        
max                                            27.101473                        

       Adjusted savings: natural resources depletion (% of GNI) - NY.ADJ.DRES.GN.ZS  \
count                                        3094.000000                              
mean                                            3.812051                              
std                                             6.958003                              
min                                             0.000000                              
25%                                             0.078542                              
50%                                             0.878741                              
75%                                             4.656705                              
max                                            71.290764                              

       Adjusted savings: net forest depletion (% 

## <b>C.7. <span style='color:#6A9AB0'>Aggregation</span></b>

Agregasi data merupakan serangkaian proses yang dilakukan untuk melakukan pengumpulan data dan dinyatakan dalam bentuk ringkasan sebagai contoh menghitung total yang dikelompokkan berdasarkan kategori tertentu, dsb.

Untuk melakukan agregasi gunakan sintaks berikut

<p align="center"><code><b>DataFrame.groupby(by, as_index).agg(kolom_hasil_aggregasi = (kolom_dihitung, jenis_agregasi))</b></code></p>

Jenis agregasi :
* sum = Menghitung total data
* mean = Menghitung rataan data
* max = Menghitung nilai maksimum data
* min = Menghitung nilai minimum data
* nunique = Menghitung jumlah data unik data
* count = Menghitung banyak data

###<b><span style='color:#55679C'>Quest C.7.1</span></b>

<i>Pada DataFrame <code>data_world_sust</code>, hitung rata-rata GDP tiap Benua pada seluruh tahun!</i>

In [94]:
# Answer C.7.1
rerata_GDP = data_world_sust.groupby(by = 'Continent', as_index= False).agg(rerata_GDP = ('GDP per capita (current US$) - NY.GDP.PCAP.CD', 'mean'))
display(rerata_GDP)

Continent    rerata_GDP
0         Africa   2190.614397
1           Asia  11860.569243
2         Europe  26220.479103
3  North America  11943.223415
4        Oceania  14645.038023
5  South America   6368.159148

###<b><span style='color:#55679C'>Quest C.7.2</span></b>

<i>Pada DataFrame <code>data_world_sust</code>, dengan cara agregasi buatkan 2 insight lainnya yang menurut kamu perlu dianalisa sertakan juga insight yang kamu peroleh!</i>

In [95]:
# Answer C.7.2.1
avg_renewable_energy = data_world_sust.groupby('Continent', as_index =False).agg(avg_renewable_energy= ('Renewable energy consumption (% of total final energy consumption) - EG.FEC.RNEW.ZS', 'mean'))

print("Rata-rata penggunaan Energi Terbarukan per Benua:")
for index, row in avg_renewable_energy.iterrows():
  display(f"{row['Continent']}: {row['avg_renewable_energy']:.2f}%")

Rata-rata penggunaan Energi Terbarukan per Benua:


'Africa: 61.45%'

'Asia: 21.31%'

'Europe: 18.77%'

'North America: 23.07%'

'Oceania: 28.27%'

'South America: 29.87%'

Insight:

Afrika merupakan benua yang memiliki angka rerata penggunaan enegeri terbarukan terbesar. sementara Asia merupakan benua dengan rerata terkecil. Semakin besar angkanya maka semakin banyak pula penggunaan energi terbarukan pada benua tersebut.

In [96]:
# Answer C.7.2.2
avg_gov_exp = data_world_sust.groupby('Continent', as_index = False).agg(
    rerata_pengeluangan_pemerintah = ('General government final consumption expenditure (% of GDP) - NE.CON.GOVT.ZS', 'mean')
)

display(avg_gov_exp)

Continent  rerata_pengeluangan_pemerintah
0         Africa                       15.018418
1           Asia                       15.421302
2         Europe                       19.287896
3  North America                       15.308493
4        Oceania                       18.516184
5  South America                       13.799687

Insight :

Europe merupakan Benua yang memiliki nilai Pengeluaran terbesar yaitu sebesar 19.287896. sementara untuk Afrika, Asia, dan North Amerika berada pada kisaran angka 15 juta, sedangkan South Amerika memiliki nilai pengeluaran terkecil yaitu sebesar 13.799687

---

<br>
<a href="https://www.linkedin.com/in/Amalihnm/"><img src="https://img.shields.io/badge/-© 2024 Amalihnm-417DAC?style=for-the-badge&logoColor=white"/></a>